pyfr/plugins/tavg.py

```diff
@@ -134,19 +165,75 @@ def _eval_acc_exprs(self, intg):
         # Stack up the expressions for each element type and return
         return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
 
-    def _eval_fun_exprs(self, intg, accex):
+    def _eval_fun_exprs_var(self, dev, accex):
+        dfexpr = []
         exprs = []
+        eps = self.eps
+        an = self.anames
+        devs = [d.swapaxes(0, 1) for d in dev]
 
         # Iterate over each element type our averaging region
         for avals in accex:
+            df = []
+            
             # Prepare the substitution dictionary
             subs = dict(zip(self.anames, avals.swapaxes(0, 1)))
-
+            
+            # Evaluate functional expressions
             exprs.append([npeval(v, subs) for v in self.fexprs])
 
+            for idx, avar in enumerate(self.anames):
+                # Calculate step size 
+                h = np.zeros_like(avals.swapaxes(0, 1), dtype=np.float64) 
+                h[idx] = np.sqrt(eps) * avals.swapaxes(0, 1)[idx]
+                h[idx][abs(h[idx]) < eps] = np.sqrt(eps)
+
+                # Prepare the substitution dictionary for step
+                subsh = dict(zip(an, avals.swapaxes(0, 1) + h)) 
+
+                # Calculate derivatives for functional averages
+                df.append([(npeval(v, subsh) - exprs[-1][i]) / h[idx]
+                             for i, v in enumerate(self.fexprs)])
+                            
+            dfexpr.append(np.stack(df))
+        
+        # Multiply by variance and take RMS value
+        fv = [np.sqrt(np.sum(df**2 * sd[:, None, ...]**2, axis = 0))
+                .swapaxes(0,1) for df, sd in zip(dfexpr, devs)]
+        
         # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
-
+        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs], fv
+    
+    def _acc_avg_var(self, intg, currex, doaccum, dowrite):  
+        prevex, vaccex, accex = self.prevex, self.vaccex, self.accex  
+                
+        # Weights for online variance and average
+        Wmp1mpn = intg.tcurr - self.prevt
+        W1mpn = intg.tcurr - self.tstart_acc
+        W1m = W1mpn - Wmp1mpn
+
+        # If initialising or first step
+        if self.tstart_acc == self.prevt:
+            # Iterate over element type
+            for v, p, c in zip(vaccex, prevex, currex):
+
+                # Accumulate variance
+                v += Wmp1mpn*(c**2 + p**2) - 0.5*Wmp1mpn*(p + c)**2
+
+        # After first step
+        else:
+            # Iterate over element type
+            for v, a, p, c in zip(vaccex, accex, prevex, currex):
+
+                # Accumulate variance
+                v += (Wmp1mpn*(c**2 + p**2) - 0.5*Wmp1mpn*(p + c)**2 + 
+                     ((W1m / (2 * Wmp1mpn * W1mpn)) * 
+                     (Wmp1mpn * a/W1m - (c + p) * Wmp1mpn)**2))
```

Comment:
Does this expression simplify if you use the new value of `accex`?  So accumulate the average first then the variance.